### PR TITLE
Allow generation of possibly invalid modules

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit takes a technique the spidermonkey folks are doing which is
to use the fuzz input as the raw body of a function. Currently many
fuzzers use the raw fuzz input as the entire wasm module, but that makes
it difficult to structure a valid wasm module because of section sizes
and such. The hope is that by simply using arbitrary fuzz input for the
instruction stream we can test more edge cases in the wasm decoder.

The changes here are to, for each function, use the fuzz input to
indicate whether a valid instruction stream is generated or a random one
is used. Modules now have a flag indicating whether they're guaranteed
valid modules (only procedurally generated functions).

One caveat is that for not-guaranteed-valid modules the
`ensure_termination` method no longer works, so wasm runtimes will need
some other method of cancelling execution while it's going.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I initially started making this a dedicated fuzz target in wasmtime, but ended up realizing it would fit a lot better here to take advantage of function types, imports, other functions, memories, etc.

I'm not sure if this is the best design, though? It may be more flexible to have a separate `MaybeInvalidModule` so `Module` is always valid.

**Comment by alexcrichton** [comment]
Hm actually looking at the fuzz failures, I'm gonna go ahead and add `MaybeInvalidModule` and it'll just naturally not have anything relation to termination too.

**Comment by fitzgen** [comment]
> Hm actually looking at the fuzz failures, I'm gonna go ahead and add `MaybeInvalidModule` and it'll just naturally not have anything relation to termination too.

Yeah this sounds like the right approach.

File: src/terminate.rs [hunk 1]
```diff
@@ -37,13 +37,17 @@ impl Module {
...
+                Instructions::Arbitrary(_) => continue,
+            };
+            let mut new_insts = Vec::with_capacity(instrs.len() * 2);
 
             // Check fuel at the start of functions to deal with infinite
             // recursion.
             check_fuel(&mut new_insts);
 
-            for inst in mem::replace(&mut code.instructions, vec![]) {
+            for inst in mem::replace(instrs, vec![]) {
```
Comment (fitzgen) [inline src/terminate.rs:50]: This can't ever be called on `MaybeInvalidModule`s, right? And therefore we shouldn't ever see `Instructions::Arbitrary` here, right? If I understand that correctly, then lets assert that here, rather than `continue`.

File: src/lib.rs [hunk 2]
```diff
@@ -689,10 +717,15 @@ impl Module {
         u: &mut Unstructured,
         ty: &FuncType,
         allocs: &mut CodeBuilderAllocations,
+        allow_invalid: bool,
     ) -> Result<Code> {
         let locals = self.arbitrary_locals(u)?;
         let builder = allocs.builder(ty, &locals);
-        let instructions = builder.arbitrary(u, self)?;
+        let instructions = if allow_invalid && u.arbitrary().unwrap_or(false) {
+            Instructions::Arbitrary(u.arbitrary()?)
```
Comment (fitzgen) [inline src/lib.rs:725]: We can do better for `Vec<u8>` than the `Vec<A: Arbitrary>` implementation does by default (can't wait for specialization...). Basically, for "raw" byte buffers, I don't think it make sense to do the `keep_going` thing that `Vec<A: Arbitrary>` does.

```rust
fn arbitrary_vec_u8(u: &mut Unstructured) -> Vec<u8> {
    let size = u.arbitrary_len::<u8>();
    u.get_bytes(size).unwrap().to_vec()
}

// ...

let instructions = if allow_invalid && u.arbitrary().unwrap_or(false) {
    Instructions::Arbitrary(arbitrary_vec_u8(u))
} else {
    // same as before...
};
```

**Comment by alexcrichton** [comment]
Updated!
